lib/rubocop/cop/rspec/empty_example_group.rb

```diff
@@ -60,15 +60,28 @@ module RSpec
       class EmptyExampleGroup < Cop
         MSG = 'Empty example group detected.'
 
-        def_node_search :contains_example?, <<-PATTERN
+        def_node_matcher :contains_example?, <<-PATTERN
           {
-            #{(Examples::ALL + Includes::ALL).send_pattern}
-            (send _ #custom_include? ...)
+            #{Examples::ALL.block_pattern}
+            #{ExampleGroups::ALL.block_pattern}
+            #{Includes::ALL.send_pattern}
+            (send nil? #custom_include? ...)
+            (block (send _ :each) _ #contains_example?)
```

Comment:
So this is a recursive matcher?

Reply:
`:each` is not the only way to define examples in an iterator (although it might be the most common), so I'm not sure why we should explicitly list it?

Reply:
> :each is not the only way to define examples in an iterator

However, if we will relax this, the following becomes possible:
```ruby
RSpec.describe Foo do
  before do
    it { is_expected.to be_valid }
  end
end
```

Do you have some other iterator methods that can be legitimately used instead of `each` to define several examples?
I can think of:
```ruby
[:a, :b].each.with_index do |o, i|
  it "#{o} #{i}" do
    # ...
  end
end
```
and similarly `each.with_object`.

I have little to no idea why would anyone use that though.

Reply:
Added support for `each.with_object`/`each_with_object`/`each.with_index`/`each_with_index`.

Reply:
Very good point indeed.

I don't have an idea what this list should include though.
Hooks are kind of obvious. With hook aliases configuration (e.g. `setup do`, `tearup do` etc) it should become easier to detect examples that are defined in improper blocks, and that shouldn't be counted towards the number of examples in an example group.
Is that it?

I remember on one of my previous projects we've used some custom block that was wrapping examples:
```ruby
RSpec.describe SomeController do
  with_permission(:read) do
    it { ... }
  end
end
```

With this change, the root example group would be detected as an empty example group.
Even though I don't personally agree it's the best approach to the problem, e.g.:
```ruby
RSpec.describe SomeController do
  context permission: :read do
    it { ... }
  end
end
```
it doesn't mean that it's an illegitimate usage.

Reply:
Let me quickly limit it to hooks, as I think it's the major issue.
We'll see if someone notices this is insufficient.

Reply:
Done.